xcm/src/v0/multi_location.rs

```diff
@@ -44,77 +44,423 @@ use parity_scale_codec::{self, Decode, Encode};
-	X4(Junction, Junction, Junction, Junction),
-	/// A relative path comprising 5 junctions.
-	X5(Junction, Junction, Junction, Junction, Junction),
-	/// A relative path comprising 6 junctions.
-	X6(Junction, Junction, Junction, Junction, Junction, Junction),
-	/// A relative path comprising 7 junctions.
-	X7(Junction, Junction, Junction, Junction, Junction, Junction, Junction),
-	/// A relative path comprising 8 junctions.
-	X8(Junction, Junction, Junction, Junction, Junction, Junction, Junction, Junction),
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Debug)]
+pub struct MultiLocation {
+	parents: u8,
+	interior: Junctions,
+}
+
+impl Default for MultiLocation {
+	fn default() -> Self {
+		Self::empty()
+	}
+}
+
+impl Encode for MultiLocation {
+	fn encode_to<T: Output + ?Sized>(&self, dest: &mut T) {
+		// 255 items max, but we cannot return an error, so we truncate silently.
+		let parents = self
+			.parents
+			.min(MAX_MULTILOCATION_LENGTH.saturating_sub(self.interior.len()) as u8);
+		let count = parents.saturating_add(self.interior.len() as u8);
+		dest.push_byte(count);
+		for _ in 0..parents {
+			dest.push_byte(0);
+		}
+		for j in &self.interior {
+			j.encode_to(dest)
+		}
+	}
+}
+
+impl Decode for MultiLocation {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, parity_scale_codec::Error> {
+		let count = input.read_byte()?;
+		let mut parents = 0u8;
+		let mut interior = Junctions::Null;
+		for i in 0..count {
+			let b = input.read_byte()?;
+			if b == 0 {
+				if i > parents {
+					// this would imply that there has been at least one non-parent junction, which
+					// further implies that this is non-canonical. we bail.
+					return Err(parity_scale_codec::Error::from("Parent after non-Parent forbidden"))
+				}
+				parents = parents.saturating_add(1);
+			} else {
+				let mut bytes = [b].to_vec();
+				loop {
+					if let Ok(junction) = Junction::decode(&mut &bytes[..]) {
+						interior = interior
+							.pushed_with(junction)
+							.map_err(|_| parity_scale_codec::Error::from("Interior too long"))?;
+						break
+					} else {
+						// a single byte may not be enough to decode a `Junction`, so keep reading
+						// the next byte until a `Junction` can successfully be decoded
+						bytes.push(input.read_byte()?);
+					}
+				}
+			}
+		}
+		Ok(MultiLocation { parents, interior })
+	}
 }
 
 /// Maximum number of junctions a `MultiLocation` can contain.
-pub const MAX_MULTILOCATION_LENGTH: usize = 8;
+pub const MAX_MULTILOCATION_LENGTH: usize = 255;
+
+impl MultiLocation {
+	/// Creates a new `MultiLocation`, ensuring that the length of it does not exceed the maximum,
+	/// otherwise returns `Err`.
+	pub fn new(parents: u8, junctions: Junctions) -> result::Result<MultiLocation, ()> {
+		if parents as usize + junctions.len() > MAX_MULTILOCATION_LENGTH {
+			return Err(())
+		}
+		Ok(MultiLocation { parents, interior: junctions })
+	}
+
+	/// Creates a new `MultiLocation` with 0 parents and a `Null` interior.
```

Comment:
```suggestion
	/// Creates a new `MultiLocation` with 0 parents and a `Null` interior.
	///
	/// This can be interpreted as the 'current consensus system'. 
	
```